proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+versatile tool for improving performance in a variety of workloads that suffer
+from divergent execution or data access.
+
+Similar to `TraceRay` and `CallShader`, the behavior of the actual reorder
+operation at any reorder point is implementation specific. An implementation
+may, for instance, not reorder at all, reorder within some local domain
+(threads on a local processor), reorder only within a certain time window,
+reorder globally across the entire GPU and the entire dispatch, or any
+variation thereof. The general target is for implementations to make a _best
+effort_ to extract as much coherence as possible, while keeping the overhead
+of reordering low enough to be practical for use in typical raytracing
+scenarios.
+
+---
+
+### Divergence behavior
+
+An implementation may only reorder threads whose control flow arrives at a
+reorder point. Threads that do not arrive at a reorder point are guaranteed to
+not migrate, even if neighboring threads in the same wave do migrate.
+
+From the physical wave's point of view, an implementation may remove, replace,
+or do nothing with threads that arrive at a reorder point. It is also legal
+for an implementation to replace terminated threads in a wave if any of its
+threads arrive at a reorder point. This may indirectly affect threads that
+conditionally did not invoke a reorder point, as illustrated in the following
+code snippet:
+
+```C++
+int MyFunc(int coherenceCoord)
+{ 
+    int A = WaveActiveBallot(true); 
+    if (WaveIsFirstLane()) 
+        ReorderThread(coherenceCoord, 32); 
+    int B = WaveActiveBallot(true); 
+    return A - B; 
+}
+```
+
+In this example, a number of different things could happen:
+- If the implementation does not honor `ReorderThread` at all, the function
+will most likely return zero, as the set of threads before and after the
+conditional reorder would be the same. 
+- If the implementation reorders threads invoking `ReorderThread` but does not
+replace them, B will likely be less than A for threads not invoking
+`ReorderThread`, while the reordered threads will likely resume execution with
+a newly formed full wave, thereby obtaining `A <= B`.
+- If the implementation replaces threads in a wave, the threads not
+participating in the reorder may possibly be joined by more threads than were
+removed from the wave, again observing `A <= B`.
+
+---
+
+### Memory coherence and visibility
+
+Due to the existence of non-coherent caches on most modern GPUs, an
+application must take special care when communicating information across
+reorder points through memory (UAVs).
+
+Specifically, if UAV stores or atomics are performed on one side of a reorder
+point, and on the other side the data is read via non-atomic UAV reads, the
+following is required:
+- The UAV must be declared `[globallycoherent]`.
+- The UAV writer must issue a `DeviceMemoryBarrier` between the write and the
+reorder point.
+
+## Separation of ReorderThread and HitObject::Invoke
```

Comment:
TraceRayReorderInvoke is similar in spirit to the attributes.

The first point is not clear to us. It is already the case that the whole wave doesn't have to wait for all rays to finish at any reorder point.

We are not quite sure why the compiler is not able to move the Sample call after ReorderThread. There are no side effects in ReorderThread and TraceRay was already a reorder point, meaning that restrictions on memory visibility apply already. A sampled texture is also read-only and is not affected by side-effects.

Our stance is that it should be easy for an implementation to merge 2-3 calls into one if it is deemed beneficial, without adding further API surface. Any shader code in between the calls is typically not affected by any side effects that may occur in TraceRay or Invoke. The only exception would be RW UAVs that are globally coherent and are accessed in between the calls. However, it is always possible to move code across ReorderThread since it does not cause or observe side effects.

Reply:
In last week's meeting we agreed to no monolithic API, but would make adjustments to the spec for clarity. 
Options going forward:

- In spec we could say calling the three functions is the same as calling trace ray semantically
- In spec explicitly call out that compiler and driver might merge these functions, and putting anything in between the function calls might prevent that.
- Have example code as a helper function. In the example code that is traceray, reorder, invoke, and the implementation does those three in a row, and it's just a sample code for the case where the app wants
- Add a helper library to compiler, and ship some helper functions as a hlsl header

@rdrabins given the above options, which would intel prefer?

Reply:
From the options above, the last one seems most preferable - a function we can reliably pattern match and "officially sanctioned".